pkg/util/clusterdata/clusterdata.go

```diff
@@ -23,7 +24,14 @@ import (
 // OpenShiftClusterEnricher must update the cluster object with
 // data received from API server
 type OpenShiftClusterEnricher interface {
-	Enrich(ctx context.Context, ocs ...*api.OpenShiftCluster)
+	Enrich(ctx context.Context, docs ...*api.OpenShiftClusterDocument)
```

Comment:
If possible, it would be nice to have implemented the single interface and multiple implementations, which can be chained. 

So if in the future we would like to have in-memory frontend catch it would be just "plug and play"?
And I think implementation should be called something like "DatabaseCache" as it is not in-memory cache and we should state it somewhere.

Reply:
I initially was not planning to introduce another method/interface and make `cachingEnricher` to be just a wrapper around any other thing that implements the `OpenShiftClusterEnricher` interface. Something like:

```go
func (e *cachingEnricher) Enrich(ctx context.Context, docs ...*api.OpenShiftClusterDocument) {
	now := e.now().UTC()
	filteredDocs := make([]*api.OpenShiftClusterDocument, 0, len(docs))

	for _, doc := range docs {
		if doc.LastEnrichment == nil || now.Sub(*doc.LastEnrichment) > e.cacheTTL {
			filteredDocs = append(filteredDocs, doc)
		}
	}

	missCount := len(filteredDocs)
	hitCount := len(docs) - len(filteredDocs)

	e.m.EmitGauge("enricher.cache.hit.count", int64(hitCount), nil)
	e.m.EmitGauge("enricher.cache.miss.count", int64(missCount), nil)

	if missCount == 0 {
		return nil
	}

	e.inner.Enrich(ctx, filteredDocs...)

	for _, doc := range filteredDocs {
		doc.LastEnrichment = &now

		_, err := e.db.OpenShiftClusters.Update(ctx, doc)
		if err != nil {
			e.log.Error(err)
			e.m.EmitGauge("enricher.cache.miss.errors", 1, nil)
		}
	}
}


```

It was beautiful until I realised that in `_putOrPatchOpenShiftCluster` we will be doing two db `Update` calls. It was not nice for several reasons including:

* Extra, unnecessary DB write
* It was easy to break: move the enrichment into a different place (after setting a non-terminal state) and after the first save the backend will start trying to dequeue the document and work on it while front-end is still processing it.